### PR TITLE
feat: add an opt-in linter job to the main workflow

### PR DESCRIPTION
**Pull request by darkgl0w**
Hello.

Following https://github.com/fastify/fastify-swagger/pull/531#pullrequestreview-860350315, this PR aims to :
- add an optional linter job to the main reusable workflow
- add documentation on how to enable the linter job

#### Checklist

- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: .github/workflows/plugins-ci-swagger.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+name: Plugin CI - Swagger
```
Comment (mcollina) [inline .github/workflows/plugins-ci-swagger.yml:1]: I would not add a plugin specific workflow in here

**Review by mcollina — CHANGES_REQUESTED** [review]

File: .github/workflows/plugins-ci-swagger.yml (hunk 1 above)
Comment (darkgl0w) [reply .github/workflows/plugins-ci-swagger.yml:1]: The plugin needs a step in which it prepares some additional files needed to run the test suite.

@mcollina > Should I remove it and change the npm scripts in `fastify-swagger` repo ?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Fdawgs** [comment]
This could be added to the existing workflows as a job, similar to what we had for the now removed coveralls input, rather than as a separate workflow.

**Comment by Eomm** [comment]
> This could be added to the existing workflows as a job, similar to what we had for the now removed coveralls input, rather than as a separate workflow?

I agree with both solutions.

Keeping the lint into a dedicated file is fine, but we should add it here on the ci* files like https://github.com/fastify/workflows/blob/main/.github/workflows/plugins-ci.yml

```
jobs:
  linter:
    uses: fastify/workflows/.github/workflows/linter-ci.yml@main
  test:
    runs-on: ${{ matrix.os }}

    strategy:
   ....
```

Note the `main`: if we write `v2` it wont work to test it.
We will set `v2` before merging.

**Comment by Fdawgs** [comment]
> > This could be added to the existing workflows as a job, similar to what we had for the now removed coveralls input, rather than as a separate workflow?
> 
> I agree with both solutions.
> 
> Keeping the lint into a dedicated file is fine, but we should add it here on the ci* files like https://github.com/fastify/workflows/blob/main/.github/workflows/plugins-ci.yml
> 
> ```
> jobs:
>   linter:
>     uses: fastify/workflows/.github/workflows/linter-ci.yml@main
>   test:
>     runs-on: ${{ matrix.os }}
> 
>     strategy:
>    ....
> ```
> 
> Note the `main`: if we write `v2` it wont work to test it. We will set `v2` before merging.

Unfortunately we [can't reference reusable workflows within reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations) 😞

**Comment by Eomm** [comment]
Oh, fine then. Thanks for pointing out.

So I think integrating it into the main plaugin-ci-* is more "comfortable" for the plugins' configuration and we are forcing a unified and optimized standard across all the repos.

**Comment by darkgl0w** [comment]
Hmmmm 🤔 

Right we can't use reusable workflows within reusable workflows 😢 🤯 

I will make the linter a part of the plugins-ci workflow (note that this will certainly require some adjustments in repos where it is deployed)

Here are some thoughts :
- we should harmonize the npm scripts among the repositories with required scripts, e.g.: https://github.com/fastify/workflows/issues/9#issuecomment-1019839890
- we should call `npm run ci:test` (or whatever name containing at least the 'ci' reference, `test:ci`, `ci:test`, `ci`...) instead of `npm test` inside the `plugins-ci` workflows
- or we can split the reusable workflows in pieces and make use of the [starter workflows](https://docs.github.com/en/actions/using-workflows/creating-starter-workflows-for-your-organization)

Here is an example of what can become a starter workflow :
```yml
name: CI workflow

on:
  push:
    paths-ignore:
      - 'docs/**'
      - '*.md'
  pull_request:
    paths-ignore:
      - 'docs/**'
      - '*.md'

jobs:
  # runs first and if errors are detected it doesn't waste resources because it will stop the dependent workflows
  linter:
    uses: darkgl0w/workflows/.github/workflows/ci-linter.yml@linter-ci

  # the test runner (currently node 10, 12, 14, 16) - when we drop a node version we remove it in one file only
  test:
    needs: linter
    uses: darkgl0w/workflows/.github/workflows/plugins-ci.yml@linter-ci

  # the test runner (only node current branch 17 - and we just need to change the version inside this workflow)
  experimental:
    needs: linter
    uses: darkgl0w/workflows/.github/workflows/plugins-ci-node-current.yml@linter-ci
```
https://github.com/darkgl0w/fastify-swagger/actions/runs/1738731030

cc @Eomm  @Fdawgs WDYT ?

**Comment by Fdawgs** [comment]
> This could be added to the existing workflows as a job, similar to what we had for the now removed coveralls input, rather than as a separate workflow?

@darkgl0w, i'd probably add it as an input which is off by default. This is so if certain repos don't need/want linting then we can disable it:

```yml
on:
  workflow_call:
    inputs:
      lint:
        description: 'Set to true to run linting scripts.'
        required: false
        type: boolean

jobs:
  lint:
    if: inputs.lint == true
     # rest of job goes here
```

Then it would be called like so:

```yml
name: CI workflow

on: [push, pull_request]
jobs:
  test:
    uses: darkgl0w/workflows/.github/workflows/plugins-ci.yml@main
    with:
       lint: true
```

**Comment by darkgl0w** [comment]
Gotcha !
I will fire some sample run on my test repo to see if the `needs` requirement will cause trouble

**Comment by darkgl0w** [comment]
@Fdawgs > I ran some sample runs following your suggestions and this is what I've come up with due to the limitations of GitHub Actions workflows :

```yml
jobs:
  test:
    needs: linter
    # workaround to always run this job
    if: ${{ always() }}
    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        node-version: [10, 12, 14, 16]
        os: [macos-latest, ubuntu-latest, windows-latest]
    steps:
      # workaround to fail if linter fail
      - name: Stop on linter job failure
        if: ${{ needs.linter.result == 'failure' }}
        run: exit 1

      - uses: actions/checkout@v2

      - name: Use Node.js
        uses: actions/setup-node@v2
        with:
          node-version: ${{ matrix.node-version }}

      - name: Install Dependencies
        run: npm i --ignore-scripts

      - name: Run Tests
        run: npm test
```
Some thoughts :
- it works despite the limitations of GitHub Actions workflows 👍 
- it will start test jobs even though the linter job fails (wasting resources until it hits the `Stop on linter job failure` step) 👎 
- I find this a little less flexible than using workflow templates with starter workflows 🤔 

for reference - workflow CI runs:
[`with lint: true` and code linting errors](https://github.com/darkgl0w/fastify-swagger/actions/runs/1739321564)
[`with lint: true` and no code linting errors](https://github.com/darkgl0w/fastify-swagger/actions/runs/1739196558)
[`with lint: false`](https://github.com/darkgl0w/fastify-swagger/actions/runs/1739211001)

**Comment by Fdawgs** [comment]
Cheers @darkgl0w, I've been busy finishing up some work pieces today, but will have a look tomorrow.

**Comment by darkgl0w** [comment]
> Do you know why this is happening?

This is a limitations that comes with the conditional launch of the linter job and it's requirement to launch the test job.

In order to be able to bypass the need of the linter job by the test job when the `lint` input is set to false we resort to a workaround (`if: ${{ always() }}`) and we add an extra check that fail the test job when there are linting errors 
```yml
      # workaround to fail if linter fail
      - name: Stop on linter job failure
        if: ${{ needs.linter.result == 'failure' }}
        run: exit 1
```
With the linter included in the main workflow behind an optional activation input that's the best we can do to reduce the waste of resources.

The other solution would be to split the jobs and to use GitHub Starter workflows. (cf. https://github.com/fastify/workflows/pull/8#issuecomment-1019861437, this is an example of what we can achieve by combining reusable workflows and starter workflows across the org repos).

Both have their pros and their cons.

**Comment by Eomm** [comment]
Thanks for explaining.
I would keep it simple:

- the fastify org is the first citizen of this workflow
- there are not repositories without the linter

So I would go for the non-skippable lint step (this should solve the need for this `Stop on linter job failure` step)

We could add an input option to set the `if-present` arg to `true` in the case we don't have a `lint` script into the `package.json`
https://docs.npmjs.com/cli/v8/commands/npm-run-script#if-present

**Comment by darkgl0w** [comment]
@Eomm > applied your suggestion and with `npm run lint --if-present` the linter job will emit an `exit 0` command only if the lint scripts does not exist ^^

[linting issues](https://github.com/darkgl0w/fastify-swagger/actions/runs/1741590562)
[no linting issues](https://github.com/darkgl0w/fastify-swagger/actions/runs/1741598615)
[no lint script](https://github.com/darkgl0w/fastify-swagger/actions/runs/1741603297)
[normal run](https://github.com/darkgl0w/fastify-swagger/actions/runs/1741628597)

File: .github/workflows/plugins-ci.yml [hunk 2]
```diff
@@ -4,25 +4,34 @@ on:
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Use Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: 16
+      - name: Install Dependencies
+        run: npm i --ignore-scripts
+      - name: Lint code
+        run: npm run lint --if-present
```
Comment (Eomm) [inline .github/workflows/plugins-ci.yml:18]: the parameter needs to be added

```suggestion
        run: npm run lint --if-present=${input.lint}
```

File: README.md [hunk 3]
```diff
@@ -32,6 +32,32 @@ jobs:
...
+  pull_request:
+    paths-ignore:
+      - 'docs/**'
+      - '*.md'
+
+jobs:
+  call-reuseable-workflow:
+    uses: fastify/workflows/.github/workflows/plugins-ci.yml@v2
+    with:
+      lint: true
```
Comment (Eomm) [inline README.md:58]:
```suggestion
      lint: false # default `true`
```

File: README.md [hunk 4]
```diff
@@ -32,6 +32,32 @@ jobs:
...
+    paths-ignore:
+      - 'docs/**'
+      - '*.md'
+  pull_request:
+    paths-ignore:
+      - 'docs/**'
+      - '*.md'
+
+jobs:
+  call-reuseable-workflow:
```
Comment (Eomm) [inline README.md:55]: easier copy-paste

```suggestion
  test:
```

File: .github/workflows/plugins-ci.yml (hunk 2 above)
Comment (Fdawgs) [reply .github/workflows/plugins-ci.yml:18]: The `lint` input has been removed, so don't think this is needed

**Comment by Fdawgs** [comment]
> https://docs.npmjs.com/cli/v8/commands/npm-run-script#if-present

Had completely forgotten this existed, this makes it so much easier.

File: README.md (hunk 3 above)
Comment (Fdawgs) [reply README.md:58]: The example needs updating to remove the `lint` input usage, as it's no longer in place in this PR.

**Comment by darkgl0w** [comment]
I have found a better solution directly inside the Github actions :P
Will push in a few seconds ^^

**Comment by darkgl0w** [comment]
With this lint is really optional and the waste of resources is minimal 😃

File: README.md (hunk 3 above)
Comment (darkgl0w) [reply README.md:58]: added back the lint input in the last iteration.

File: .github/workflows/plugins-ci.yml (hunk 2 above)
Comment (darkgl0w) [reply .github/workflows/plugins-ci.yml]: added back the lint input in the last iteration.

**Review by Fdawgs — APPROVED** [review]
...Where's the chef's kiss emoji? 👨‍🍳
LGTM

**Comment by darkgl0w** [comment]
I think that we can push those workflows even further and choose to exclude/include some node/os versions.

File: README.md [hunk 5]
```diff
@@ -32,6 +32,32 @@ jobs:
 
 Included in this repo is a [basic workflow](.github/workflows/plugins-ci.yml) for use across the majority of plugins, as well as variants with service containers.
 
+### Enable workflow Linter job
+
+By setting the `lint` value to `true` when using the [basic workflow](.github/workflows/plugins-ci.yml) the CI will first run the linter job.
```
Comment (Eomm) [inline README.md:37]:
```suggestion
By setting the `lint` option to `true` when using the [basic workflow](.github/workflows/plugins-ci.yml) the CI will first run the linter job once.

```

**Review by Eomm — APPROVED** [review]

**Pull request merged**